### DISCUSSION TITLE
serial developer documentation

### DISCUSSION THREAD

**Discussion by epragalakis** [meshtastic/Discussions#138] [Firmware]
Hello, 
I'm writing tiny modules to interact with the serial on a RAK device.
So far I was not able to find any documentation other than this: https://meshtastic.org/docs/configuration/module/serial/

I'm able to read data but I'm wondering on some technical details on how the whole thing works.

From a small log sample the structure of each log looks like this: `DEBUG|INFO| possibly more? hh:mm:ss SOME_CODE_3_DIGIT_NUMBER [Power | DeviceTelemetry | Router | GPS | possibly more] MSG`
is there any documentation on the structure of a log? (eg: what is that 3 digit number, how do categories translate to functionality like connections metadata or messages, why do certain logs miss a category, how does log severity translate to the log type etc)

Is it possible to only get a certain category or log severity of logs? (eg: I only care about getting messages and node connections metadata or I only want to see info or error logs)

**Comment by epragalakis** [comment]
Also Im wondering about this line: https://github.com/meshtastic/firmware/blob/master/src/SerialConsole.cpp#L75

as I understand, the client will disconnect if I dont send a message to the device (keep alive), but that has not been my experience, where I am able to read data without doing that (unless the connection drops after running it for long time, which is not something that Ive tried yet)

**Comment by epragalakis** [comment]
is there a way to send binary data from the device?

**Comment by epragalakis** [comment]
is there a way to completely disable functionality that its not needed?
(eg: my device does not have GPS, so getting logs related to that does not make sense)

**Reply by garthvh** [reply]
Set gps to not present and it won't run any of the detection code, enabled or disabled will look for hardware and you can't use or disable a device that is not there.

**Comment by barryhunter** [comment]
If you are seeing these 'DEBUG' messages, suggest you connected to the 'primary' serial connection, which is normally encapsulated via USB for taking to a host computer. 

I dont think those debug messages are intended for programmatic consumption as such, they well for developer debugging!



The 'Serial' **Module** you link to, controls a separate UART, eg so can connect it to another MCU. 

On this seperate Serial, if you want to send/recieve binary messages, probably best to wrap in proto messages. 
... ie set the the module to talk with proto messages. 



If you really are trying to interact with the primary serial connection, suggest the python client is the best reference
https://github.com/meshtastic/python
its a reasonly complete library for talking over serial (and others) to a meshtastic firmware node.
